### ISSUE TITLE
Sometimes the policy list contains double (or triple) entries

### ISSUE THREAD

**Issue by plettich** [privacyidea/privacyidea#1132]
In certain cases when viewing the policy list in the webui, the entries are doubled. This is not a javascript issue but the API itself delivers the doubled entries. The corresponding database table still contains the correct amount of entries.
This could be a race condition somewhere in [`PolicyClass.get_policies()`](https://github.com/privacyidea/privacyidea/blob/ff9427c80d4704bafff9fe4597f1ecffb230172d/privacyidea/lib/policy.py#L447).
A quick and dirty fix could be to remove the double entries before returning the list of policies.

**Comment by fredreichbier** [comment]
I have also seen this sometimes. I would guess that this is indeed a race condition in ``PolicyClass`` for which a singleton instance is shared among threads. We should probably have one instance per request (i.e. thread-local in ``flask.g``).

**Comment by fredreichbier** [comment]
Here is a script that manages to reproduce the race condition sometimes: https://gist.github.com/fredreichbier/6ad833a51eaa5152626551c7dddf6493
It requires ``requests_threads``. To run it, just create one policy ``hide_welcome`` (and just this one), and run it with
```
python script.py https://privacyidea/ adminuser adminpass
```
After some time, it prints
```
have the following policies: [u'hide_welcome', u'hide_welcome']
```

Naturally, this can only happen if we have a multi-threaded WSGI environment.

**Comment by fredreichbier** [comment]
The problem is that every process holds exactly one ``PolicyClass`` object which manages the policies for all threads. This leads to some problems:
* The problem above, in which two threads reload the policy list in parallel, which results in duplicate entries.
* Policies are not guaranteed to be static during one request! The following scenario is admittedly artificial, but possible:
   * Thread 1 starts to handle request A, starts handling policies
   * Thread 2 starts to handle request B and for some reason, reloads the policy list
   * The second half of the handling of request A operates on the new policy table

I think it would be desirable to solve both issues while not affecting performance negatively -- I guess the way to go would be
* Have a shared policy object which holds the "current" policy list and timestamp
* When a new request comes in, it checks if the shared policy list should be reloaded according to the ``PI_CHECK_RELOAD_CONFIG`` setting and if yes, does so. Reloading is done with a lock (solves the first issue). The request then clones the current policy list and holds its own copy for the complete request (solves the second issue).

As ``ConfigClass`` works similarly, we could use the sample approach there.

**Comment by fredreichbier** [comment]
This was closed by accident by my suboptimal commit message :)

**Comment by fredreichbier** [comment]
I opened #1491 which implements the approach above for the config class.

**Comment by fredreichbier** [comment]
#1491 implements the approach for the config class, but needs extensive testing: We need to ensure that requests that change configuration actually invalidate the request-local config object, so that config changes are visible to that request.

For the policy class, the case is more complicated: Many policy decorators allow to pass a ``g`` object in the ``options`` dictionary. Many tests use this feature and pass a custom policy object via the ``g`` option. So if we simply store the request-local policy object in the request-local store, this breaks a lot of tests.

Thus, we move this to the 3.1 milestone.

**Comment by fredreichbier** [comment]
This should be finally fixed by 731fc41
